File: 42-trapping-rain-water/Python/trapping-rain-water[optimal-n].py
'''
    STATUS: Time Limit Exceeded
    Time Complexity: 0(n)
    Space Complexity: 0(1)
'''

from typing import List

heights = [0, 1, 0, 2, 1, 0, 3, 1, 0, 1, 2];

class Solution:
    def get_trapped_rain_water(self, height: List[int]) -> int:
        left = max_left = max_right = total_water = 0
        right = len(height) -1

        while left < right:
            if(height[left] <= height[right]):
                if height[left] >=  max_left:
                    max_left = height[left]
                else:
                    total_water = total_water + max_left - height[left]
                left = left +1
            else:
                if height[right] >= max_right:
                    max_right = height[right]
                else:
                    total_water = total_water + max_right - height[right]
                right = right - 1

        return total_water

        
s = Solution()
# s.get_trapped_rain_water(heights)
print(s.get_trapped_rain_water(heights))

File: 42-trapping-rain-water/JavaScript/trapping-rain-water[optimal-n].js
/* 
  STATUS: Time Limit Exceeded
  Time Complexity: 0(n)
  Space Complexity: 0(1)
*/

const heightsArray = [0, 1, 0, 2, 1, 0, 3, 1, 0, 1, 2];

const getTrappedRainwater = function (heights) {
  let left = 0;
  maxLeft = maxRight = totalWater = 0;
  right = heights.length - 1;
  while (left < right) {
    if (heights[left] <= heights[right]) {
      // operate left
      if (heights[left] >= maxLeft) {
        // update leftMax
        maxLeft = heights[left];
      } else {
        // calculate water
        totalWater += maxLeft - heights[left];
      }
      left++;
    } else {
      // operate right
      if (heights[right] >= maxRight) {
        maxRight = heights[right];
      } else {
        totalWater += maxRight - heights[right];
      }
      right--;
    }
  }
  return totalWater;
};

console.log(getTrappedRainwater(heightsArray));
